controllers/statestorage.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2021 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"errors"
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/alexedwards/scs/v2"
+	"go.uber.org/zap"
+)
+
+type StateStorage struct {
+	SessionManager *scs.SessionManager
+}
+
+const (
+	letterBytes = "abcdefghijklmnopqrstuvwxyz1234567890"
+)
+
+func (storage StateStorage) VeilRealState(req *http.Request) (string, error) {
+	state := req.URL.Query().Get("state")
+	if state == "" {
+		zap.L().Error("Request has no state parameter")
+		return "", errors.New("request has no `state` parameter")
+	}
+	newState := randStringBytes(32)
```

Comment:
We could DDoS the service with this, IMHO, because everytime `/authenticate` is accessed, we add new data to the server-side session. What about using a SHA of the original state as the veil?